src/System.Management.Automation/engine/parser/ast.cs

```diff
@@ -847,6 +847,84 @@ public class ScriptBlockAst : Ast, IParameterMetadataProvider
             }
         }
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ScriptBlockAst"/> class.
+        /// This construction uses explicitly named begin/process/end/dispose blocks.
+        /// </summary>
+        /// <param name="extent">The extent of the script block.</param>
+        /// <param name="usingStatements">The list of using statments, may be null.</param>
+        /// <param name="attributes">The set of attributes for the script block.</param>
+        /// <param name="paramBlock">The ast for the param block, may be null.</param>
+        /// <param name="beginBlock">The ast for the begin block, may be null.</param>
+        /// <param name="processBlock">The ast for the process block, may be null.</param>
+        /// <param name="endBlock">The ast for the end block, may be null.</param>
+        /// <param name="cleanupBlock">The ast for the cleanup block, may be null.</param>
+        /// <param name="dynamicParamBlock">The ast for the dynamicparam block, may be null.</param>
+        /// <exception cref="PSArgumentNullException">
+        /// If <paramref name="extent"/> is null.
+        /// </exception>
+        [SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", MessageId = "param")]
+        public ScriptBlockAst(
+            IScriptExtent extent,
+            IEnumerable<UsingStatementAst> usingStatements,
+            IEnumerable<AttributeAst> attributes,
+            ParamBlockAst paramBlock,
+            NamedBlockAst beginBlock,
+            NamedBlockAst processBlock,
+            NamedBlockAst endBlock,
+            NamedBlockAst cleanupBlock,
+            NamedBlockAst dynamicParamBlock)
+            : base(extent)
+        {
+            SetUsingStatements(usingStatements);
+
+            if (attributes != null)
+            {
+                this.Attributes = new ReadOnlyCollection<AttributeAst>(attributes.ToArray());
+                SetParents(Attributes);
+            }
+            else
+            {
+                this.Attributes = s_emptyAttributeList;
+            }
+
+            if (paramBlock != null)
+            {
+                this.ParamBlock = paramBlock;
+                SetParent(paramBlock);
+            }
+
+            if (beginBlock != null)
+            {
+                this.BeginBlock = beginBlock;
+                SetParent(beginBlock);
+            }
+
+            if (processBlock != null)
+            {
+                this.ProcessBlock = processBlock;
+                SetParent(processBlock);
+            }
+
+            if (endBlock != null)
+            {
+                this.EndBlock = endBlock;
+                SetParent(endBlock);
+            }
+
+            if (cleanupBlock != null)
```

Comment:
Personally, I think we should call this new block `clean`. `begin`, `process` and `end` are all verbs, mapping to BeginProcessing, ProcessRecord, and EndProcessing. Cleanup is a noun, and thus we should use `clean`, meaning CleanResource.
But the name can be left to the end to decide, let's not change it now, otherwise the diff would be everywhere and it would be hard to review the more important functional changes.

Reply:
FWIW this is one of the reasons I initially didn't really want to change it away from `dispose` 😁 

`clean` works but yeah, we'll leave that till the end for now.